### ISSUE TITLE
vips_magickload behaves differently from vips_magickload_buffer when dealing with NEF files

### ISSUE THREAD

**Issue by lkempf** [libvips/libvips#3380]
## Bug report

**Describe the bug**
When loading a NEF file vips_magickload produces a image of the correct dimension while vips_magickload_buffer produces a image of dimension 160x120 (probably some embedded thumbnail). This small image is also produced when loading the image as a TIFF (which is what libvips identifies the image as).

**To Reproduce**
1. Use Image https://www.mannyphoto.com/D700D3/_DSC4175.NEF
2. Execute the following in python
   ```
import pyvips
img_bytes = open('./_DSC4175.NEF', 'rb').read()
img1 = pyvips.Image.magickload('./_DSC4175.NEF')
img2 = pyvips.Image.magickload_buffer(img_bytes)
print(img1.width, img2.width)
   ```
3. Output is `4284 160` not the expected `4284 4284`.

**Expected behavior**
Both methods should produce the same image.

**Actual behavior**
Loading from buffer produces a much smaller image.

**Environment**
 - OS: Arch Linux
 - Vips: vips-8.14.1

**Comment by kleisauke** [comment]
ImageMagick doesn't have a buffer recognizer for Nikon's NEF file format. Since this file format uses the standard TIFF header, it tries to load the buffer through its TIFF loader. libvips has a special path to workaround this limitation of *magick, but that does not include support for NEF. 

I had a quick attempt to add a NEF sniffer for this, see e.g. commit https://github.com/kleisauke/libvips/commit/ab5d0e356f4f6f2ca4cbca2a07c0aff7b3a8ae34, but somehow(?) this would cause a regression on the file-based loader, as it outputs `160 4284` on the above reproducer.

**Comment by lkempf** [comment]
That's not the output of the file based loaders, that's the width for both loaders. Sorry for the confusing output format.

**Comment by kleisauke** [comment]
> this would cause a regression on the file-based loader, as it outputs `160 4284` on the above reproducer.

I just reported this upstream at https://github.com/ImageMagick/ImageMagick/issues/6242.

**Comment by kleisauke** [comment]
I just opened PR #3460 for this.

**Comment by uhthomas** [comment]
I can see that https://github.com/libvips/libvips/pull/3460 did not actually include a fix for this @kleisauke. Do you have any ideas on how we can reliably differentiate between NEF and TIFF files?

**Comment by kleisauke** [comment]
@uhthomas AFAIK, the only reliable way is to get parse the make TIFF tag and check for `NIKON CORPORATION` or just `NIKON`. See for example how it's handled in [RawSpeed](https://github.com/darktable-org/rawspeed):
https://github.com/darktable-org/rawspeed/blob/75bfffedf5378a17136da18d1560d7e0c3509338/src/librawspeed/decoders/NefDecoder.cpp#L63

This would also mean that the magick sniffer needs to inspect more than 256 bytes, which would probably affect performance.

Looking at the linked issues, I think the issue is that `tiffload` has a higher priority than `magickload`.
https://github.com/libvips/libvips/blob/8d5f33174c956623c1b7eb82c43c828c0511cc22/libvips/foreign/tiffload.c#L181-L183
https://github.com/libvips/libvips/blob/8d5f33174c956623c1b7eb82c43c828c0511cc22/libvips/foreign/magick7load.c#L365-L368

One way to fix this is to prevent TIFF-based RAW images from being loaded via libtiff, see e.g. commit https://github.com/kleisauke/libvips/commit/a061854e5b553e41ba5dba9657081e41504e8a7f (this feels quite hackish :sweat_smile:).

A dedicated LibRaw/RawSpeed loader would make more sense, please subscribe to #2075 for updates regarding this.

**Comment by jcupitt** [comment]
> Looking at the linked issues, I think the issue is that tiffload has a higher priority than magickload.

Yes, the libvips TIFF loader is quite a bit quicker than the imagemagick one, so it needs to be higher priority. As you say, the best solution is probably adding a libraw loader to libvips. Someone just neds to do the work heh.

**Comment by mertalev** [comment]
We fixed the issue for now by compiling libvips without libtiff. Adding libraw support would be amazing, but the `make` tag solution sounds great too. I would prefer a minor performance hit if it meant a reliably correct output, though I obviously can't speak for other libvips users.

**Comment by zynexiz** [comment]
I experienced same issue with small images generated with vips. I have Immich installed on my server (from AUR, no docker installation), and when uploading NEF files, it generates a 160x120 instead of full size image. I tried running vips locally on my laptop, and the conversion from NEF to WebP have the same issue.

**Comment by kleisauke** [comment]
A dedicated LibRaw loader is now available on the master branch (which will be part of the upcoming 8.18 release).

This loader has a higher priority than the TIFF loader and only checks filename suffixes, so the issue no longer occurs.
https://github.com/libvips/libvips/blob/e00beb3601e632c5ea0a7df4571dee54e5ed1ff5/libvips/foreign/dcrawload.c#L407-L411

I'll close.
